### PR TITLE
Added promotion from integer to inexact to avoid overflow in `_fft_norm`

### LINKED ISSUE #18453: Large multi-channel inverse FFTs give wrong results
### Description

```python
import jax.numpy as jnp
from jax import random

key = random.PRNGKey(0)

def fun(key, length):
    size = (length,) * 3
    last3 = range(-3, 0)
    f = random.normal(key, (3,) + size)  # 3 channels
    f = jnp.fft.fftn(f, axes=last3) 

    # 1st way to do the same inverse FFT, all at once
    f_ = jnp.fft.ifftn(f, axes=last3)
    print(f_[0, 0, 0, 1])
    del f_

    # 2nd way to do the same inverse FFT, channel by channel
    f_ = jnp.stack([jnp.fft.ifftn(g) for g in f], axis=0)
    print(f_[0, 0, 0, 1])

fun(key, 890)  # prints (-0.93745315-9.771974e-08j) (-0.93745315-9.771974e-08j)
fun(key, 900)  # prints (-1235954700+99.51599j) (-1.6954112+1.9948014e-07j)
```

For 3D FFT with grid size <= 890, the two methods to do the same inverse FFT agree and give the correct results (expected to be or close to be real numbers), while for size >= 900, the 1st method gives the wrong result.

### What jax/jaxlib version are you using?

jax 0.4.19 jaxlib 0.4.19+cuda12.cudnn89

### Which accelerator(s) are you using?

GPU

### Additional system info

_No response_

### NVIDIA GPU info

```
Thu Nov  9 07:27:37 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.10              Driver Version: 535.86.10    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA H100 PCIe               Off | 00000000:19:00.0 Off |                    0 |
| N/A   40C    P0              73W / 350W |  80821MiB / 81559MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A    434548      C   ...ome/yinli/mamba/envs/jax/bin/python    80808MiB |
+---------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by Micky774**
Fixes https://github.com/google/jax/issues/18453

If a product would result in an integer overflow, now casts to `float32` to complete the calculation.

cc: @hawkinsp

### CODE REVIEW

**Comment by Micky774** [comment]
@jakevdp This PR should be a bit more sensible. Let me know what you think!

**Comment by Micky774** [comment]
cc: @pearu

File: jax/_src/numpy/fft.py [hunk 1]
```diff
@@ -32,6 +33,15 @@
 def _fft_norm(s: Array, func_name: str, norm: str) -> Array:
   if norm == "backward":
     return jnp.array(1)
+
+  # Avoid potential integer overflow
+  if s.dtype.kind in "ui":
```
Comment (pearu) [inline jax/_src/numpy/fft.py:38]:
```suggestion
  if jnp.issubdtype(s.dtype, jnp.integer)
```

File: jax/_src/numpy/fft.py [hunk 2]
```diff
@@ -32,6 +33,15 @@
...
+    MAX_LOG = 63 if config.enable_x64.value else 31
+    MAX_LOG *= np.log(2)
+    log_val = reductions.sum(ufuncs.log(s))
+    if log_val >= MAX_LOG:
+      s = s.astype(jnp.float32)
```
Comment (pearu) [inline jax/_src/numpy/fft.py:43]: This looks very expensive to just determine if the conversion to float is required.

A simpler approach would be to always convert all integer inputs to float inputs to make sure that there will be no integer overflow:
```suggestion
    s = s.astype(jnp.float32)
```
or
```suggestion
    s = s.astype(jnp.float64 if jax.config.jax_enable_x64 else jnp.float32)
```

**Review by pearu — CHANGES_REQUESTED** [review]
Added a suggestion that might be more efficient.

File: jax/_src/numpy/fft.py [hunk 3]
```diff
@@ -32,6 +33,11 @@
 def _fft_norm(s: Array, func_name: str, norm: str) -> Array:
   if norm == "backward":
     return jnp.array(1)
+
+  # Avoid potential integer overflow
+  if jnp.issubdtype(s.dtype, jnp.integer):
+    s = s.astype(jnp.float64 if config.jax_enable_x64 else jnp.float32)
+
```
Comment (jakevdp) [inline jax/_src/numpy/fft.py:40]: There's a utility for this
```python
from jax._src.numpy.util import promote_dtypes_inexact

...

s, = promote_dtypes_inexact(s)
```
You can call it regardless of the dtype of `s`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for this! Comment below.

Also, we should add a regression test in `tests/fft_test.py`

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!

File: jax/_src/numpy/fft.py [hunk 4]
```diff
@@ -26,13 +26,18 @@
 from jax._src.numpy import lax_numpy as jnp
 from jax._src.numpy import ufuncs, reductions
 from jax._src.typing import Array, ArrayLike
+from jax._src.numpy.util import promote_dtypes_inexact
```
Comment (jakevdp) [inline jax/_src/numpy/fft.py:29]: Please add this to the existing imports from `jax._src.numpy.util`, a few lines above.

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Comment by jakevdp** [comment]
Thanks! It looks like there are some conflicts with the main branch that must be resolved before we can proceed. Can you rebase (not merge) on `main`? Thanks!

**Review by jakevdp — APPROVED** [review]

**Pull request merged**